text/2017-07-23-relaxed-memory.md

```diff
@@ -0,0 +1,218 @@
+unlinked.  The unlinked location is safe to deallocate later when the epoch is advanced enough.  As
+you may guess, the correctness of EBR crucially depends on the invariant on the epoch: it is
+advanced only when old locations unlinked in the past are no longer accessible.
+
+EBR is introduced in [Keir Fraser's PhD thesis][kfraser], and later Aaron Turon introduced it
+in [his blog post on Crossbeam][aturon] to the Rust community.  Both explains EBR in details so that
+a reader can understand not only the intuition behind it but also why it is correct.  However, both
+assumed we are on the *sequentially consistent (SC) concurrency model*: every access to the memory
+from all the threads are *linearizable*, and each load reads from the most-recently stored value
+according to the linearization order.  However, unfortunately, the real-world doesn't allow us to
+enjoy the luxury of SC.  In particular, the C11/C++11/LLVM/Rust memory model is *relaxed*, or
+*weak*, allowing each load to read from stale old values.
+
+As a result, it is still unclear after two years of development whether the implementation of
+Crossbeam is correct w.r.t. the C11 memory model.  We believe that in order for Crossbeam to indeed
+serve as the crossbeam for Rust concurrency, as `java.util.concurrent` does for Java concurrency,
+Crossbeam's correctness should be formally analyzed and hopefully verified.  This RFC aims for
+providing a step towards the goal.
+
+## Non-motivation
+
+Crossbeam is the marriage of EBR and Rust's linear type system with lifetime.  However, this RFC
+does not aim for clarifying how Rust's type system helps to ensure the correctness of Crossbeam.
+
+
+
+# Detailed design
+
+In this section, we present the proposed implementation of Crossbeam, and explain why it is correct
+w.r.t. the C11 memory model.
+
+
+## Proposed implementation
+
+As discussed in Aaron Turon's original [blog post][aturon]
+and [the pull request of a previous RFC][rfc2-pr], Crossbeam guarantees that *every access to a
+location should happen before the location is deallocated*.  In order to provide this guarantee,
+Crossbeam assumes that when a pinned thread marks a location as unlinked, *the thread already
+removed all the references to the location from the memory on the thread's point of view.*; and
+Crossbeam defers the deallocation of an unlinked location until all the threads concurrently pinned
+at the time the location is unlinked, are unpinned.
+
+In order to track the concurrently pinned threads, Crossbeam utilizes the epoch as follows:
+
+- There is a local epoch `th.epoch` for each *pinned* thread `th`, and a global epoch `EPOCH`.
+
+- When a location is deallocated, it is marked with the **current** global epoch, as in the
+  following pseudocode (for now, please ignore the fences and the orderings such as `Relaxed`,
+  `Acquire`, `Release`, and `SeqCst`):
+
+  ```rust
+  fn unlink(l) {
+    'L10: atomic::fence(SeqCst);
```

Comment:
Sure. Consider your proposed implementation:

```rust
fn unlink() {
    atomic::fence(SeqCst);
    let X = epoch.load(Relaxed);
    // ...
}
```

And the alternative:

```rust
fn unlink() {
    let Y = epoch.load(Relaxed);
    // ...
}
```

Also, let's say that `G_pin` is the global epoch just after the current thread was pinned:

```rust
fn pin(th) {
    let e = EPOCH.load(Relaxed);
    th.epoch.store(e, Relaxed);
    atomic::fence(SeqCst);
    let G_pin = EPOCH.load(Relaxed);
}
```

Note that:

1. The current thread is pinned while `unlink` is called.
2. `G_pin <= X <= G_pin + 1`, because the global epoch can advance at most once while our thread is pinned.
3. `G_pin <= Y`, because `G_pin` was loaded before `Y`.
4. From 2. and 3. we can derive: `Y >= X - 1`.

The invariant is that garbage tagged with `X` can be destroyed if `G >= X + 2`.
Or, equivalently, garbage can be destroyed if `G >= X - 1 + 3`.

Since `Y + 3 >= X - 1 + 3`, we can be slightly **more conservative** and say.
Garbage can be destroyed if `G >= Y + 3`.
This is a slightly stronger inequality than `G >= X + 2`.